bigcache.go

```diff
@@ -22,6 +22,18 @@ type BigCache struct {
 	maxShardSize uint32
 }
 
+// RemoveReason is a value used to signal to the user why a particular key was removed in the OnRemove callback.
+type RemoveReason int32
+const (
+	// Expired means the key is past its LifeWindow.
+	Expired    = 1 << iota
```

Comment:
from my understanding, what your suggesting will assign the sequence of values as 0, 1, 2

However, for bitmasks, the correct sequence is 1,2,4,8,16 etc
since the binary representation of 4 is, for example, 0100, you can combine these masks with the bitwise | to store multiple bools on the same integer. Thus, Expired | Deleted actually equals 0101, which is what lets me store just one integer instead of several bools

Reply:
Mmmm while I do appreciate the optimisation, I am going to disagree and say while it is an integer optimization, the benefit of the optimisation underweighs against the idiotmatic use of `iota`. When you filed the PR, you made this statement:

> I'm still pretty new to golang, so this PR probably has a C-like odor to it

Bit shifting an `iota` is something you shouldn't really be doing in idiomatic Go as a lot of the reason for `iota` was to provide an enumerator that you don't have to manage or think about - the runtime manages it.

I'm going to defer to @janisz, but my stance is that bit shifting an `iota` is over-engineering and not idiomatic.

Reply:
> In C we always avoid doing long, sequential boolean checks using bitmaps.

I agree, but this is Go, [you can always do](https://play.golang.org/p/paDVmrAH9rC):

```go
switch reason {
  case Expired: //
  case NoSpace: // 
  case Deleted: //
 }
```

Which is both cleaner logic as well as being idiomatic. More lines of code for idiomacy is better.

Reply:
The switch is definitely cleaner- but it still requires 3x varaibles on the config instead of one, for the filter